sound/soc/sof/intel/hda-ipc.c

```diff
@@ -234,19 +234,21 @@ irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
 {
```

Comment:
please explain in your commit message how the error can happen, e.g. something like:

A race condition can happen if the first part of the next interrupt deals with stream events while the interrupt thread of the current interrupt dealing with IPC hasn't been scheduled yet.

Reply:
I previously explained but you commented "makes no sense..."
```
The rest makes no sense...
otherwise, the previously stored one may be
modified(to 0 usually) when a consequent IOC interrupt arrives.
```

Reply:
To avoid more backwards and forwards here, let me paste the whole commit message for that commit to confirm it directly here:

@plbossart can you help check if this make sense?

```
    ASoC: SOF: Intel: hda-ipc: save ipc irq_status only for IPC IRQ
    
    We are using a shared IRQ for IPC and stream IOC interrupts. A race condition
    can happen if the first part of the next interrupt deals with stream events while the
    interrupt thread of the current interrupt dealing with IPC hasn't been scheduled yet.
    
    We need to check the interrupt source before storing the status, and only save ipc
    irq_status when the interrupt source is IPC.
    
    Here add check to fix that.
```

Reply:
@keyonjie I still disagree that this check is required at all. The thread will be scheduled only it is an IPC interrupt. Why do we need to store it and check again. 
Is there a possibility of the thread getting scheduled when a stream interrupt arrives?

Reply:
> @keyonjie I still disagree that this check is required at all. The thread will be scheduled only it is an IPC interrupt. Why do we need to store it and check again.
> Is there a possibility of the thread getting scheduled when a stream interrupt arrives?

@ranj063 please check in the source code, we store it irq handler, not thread.

```
irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
{
	struct snd_sof_dev *sdev = context;
	int ret = IRQ_NONE;
	u32 irq_status = 0;

	spin_lock(&sdev->hw_lock);

	/* store status */
	irq_status = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIS);
	dev_vdbg(sdev->dev, "irq handler: irq_status:0x%x\n", irq_status);

	/* invalid message ? */
	if (irq_status == 0xffffffff)
		goto out;

	/* IPC message ? */
	if (irq_status & HDA_DSP_ADSPIS_IPC) {
		sdev->irq_status = irq_status;
		/* disable IPC interrupt */
		snd_sof_dsp_update_bits_unlocked(sdev, HDA_DSP_BAR,
						 HDA_DSP_REG_ADSPIC,
						 HDA_DSP_ADSPIC_IPC, 0);
		ret = IRQ_WAKE_THREAD;
	}

out:
	spin_unlock(&sdev->hw_lock);
	return ret;
}
```

Reply:
@keyonjie I understand. what I am asking is why do we need to check if it is an IPC irq again in the thread?

Reply:
sorry don't follow, we don't check that in the thread.

Reply:
@keyonjie I was talking about this piece of code in hda-ipc.c

irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
{
        struct snd_sof_dev *sdev = context;
        irqreturn_t ret = IRQ_NONE;
        u32 hipci;
        u32 hipcie;
        u32 hipct;
        u32 hipcte;
        u32 hipcctl;
        u32 msg;
        u32 msg_ext;

        /* here we handle IPC interrupts only */
        if (!(sdev->irq_status & HDA_DSP_ADSPIS_IPC)) **>>>>>> this is where the irq_status is used. why do we need to check this again in the thread?**
                return IRQ_NONE;
.....
}

Reply:
> @keyonjie I was talking about this piece of code in hda-ipc.c
> 
> irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
> {
> struct snd_sof_dev *sdev = context;
> irqreturn_t ret = IRQ_NONE;
> u32 hipci;
> u32 hipcie;
> u32 hipct;
> u32 hipcte;
> u32 hipcctl;
> u32 msg;
> u32 msg_ext;
> 
> ```
>     /* here we handle IPC interrupts only */
>     if (!(sdev->irq_status & HDA_DSP_ADSPIS_IPC)) **>>>>>> this is where the irq_status is used. why do we need to check this again in the thread?**
>             return IRQ_NONE;
> ```
> .....
> }

here it is the only one place we check it.

Reply:
but hda_dsp_ipc_irq_handler() could be invoked at IOC interrupt happens, our IPC share interrupt number with IOC.

Reply:
@keyonjie agreed but hda_dsp_ipc_irq_handler checks the irq_status to see if it is an IPC interrupt and returns IRQ_WAKE_THREAD only then. Why do we need to check again in the thread?

Reply:
got your point now @ranj063 , you mean we don't need the check in
hda_dsp_ipc_irq_thread(), right?

if so, yes, that's existed superfluous, let me remove them.

Reply:
@keyonjie in this case, we can remove irq_status from sdev and make it a local variable in the handler